### DISCUSSION TITLE
CI Testing Serverless Functions

### DISCUSSION THREAD

**Discussion by ryanjwilke** [vercel/vercel#4098] [General] [answered]
Hello,

I was curious if anyone had any experience with testing the serverless functions locally or in a CI service like GitHub Actions.

It's nice that everything works locally just fine, but I'd love to have a way to make sure `now dev` was always running in the background during the CI tests to ensure the API functions are working as expected as well.

Any thoughts on all this?

**Comment by coetry** [comment, marked as answer]
> I'd love to have a way to make sure now dev was always running in the background during the CI tests to ensure the API functions are working as expected as well.

Can you elaborate on this a little? Typically what we recommend for testing is to "test in prod". You can deploy your projects to a non-production branch and still see how it would run in a production environment. You can run numerous "checks" against your PRs. 

@rauchg touches upon that in his [2019 Review Post](https://rauchg.com/2020/2019-in-review#testing-the-jamstack)

**Reply by ryanjwilke** [reply]
Thanks coetry. I was able to figure out a workaround to this by publishing to a staging/preview server first to make sure the CI tests are running against the functions published in the "staging" environment. This works well, but I was ideally hoping to have a way to run the `now dev` command inside a CI build environment so all of the serverless functions could be tested as well.

Thanks for the link to the post. I'll look it over.

**Reply by ryanjwilke** [reply]
As a little additional info, this article by semaphore is roughly what I'm trying to achieve: https://semaphoreci.com/blog/serverless-cicd-hosting-semaphore-zeit

**Reply by coetry** [reply]
`now dev` just opens up a mini dev server that can be executed in any server environment. What sorts of limitations are you running into? You may have to adjust port mapping since we use port 3000 by default.

**Reply by hhimanshu** [reply]
Are there any github projects which demonstrate using `vercel dev` for local tests and also in CI?

**Comment by ctrlplusb** [comment]
For anyone who may be interested I created a package that allows unit/integration testing of your lambdas:

https://github.com/ctrlplusb/vercel-node-server

**Reply by markozxuu** [reply]
That's great ! Thank You @ctrlplusb : )
